Enable RRef timeout for tensorpipe

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39531 Enable RRef timeout for tensorpipe**
* #39530 Cleanup rref_impl

Enables RRef timeout support in TP agent by having TP agent mark
timeout errors with `makeRPCError` API. Also does some refactoring so TP agent
can print out the timeout for each future that has timed out.

Differential Revision: [D21881475](https://our.internmc.facebook.com/intern/diff/D21881475/)

rohan-varma committed:
Enable RRef timeout for tensorpipe

Enables RRef timeout support in TP agent by having TP agent mark
timeout errors with `makeRPCError` API. Also does some refactoring so TP agent
can print out the timeout for each future that has timed out.

Differential Revision: [D21881475](https://our.internmc.facebook.com/intern/diff/D21881475/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! Please get a stamp from @lw too

rohan-varma commented:
@lw Do you mind taking a look at this diff? Thanks!

osalpekar commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -23,6 +24,8 @@ const std::string kNumIdleThreads = "agent.num_idle_threads";
 const std::string kClientActiveCalls = "agent.client_active_calls";
 const std::string kServerActiveCalls = "agent.server_active_calls";
 const std::string kServerActiveAsyncCalls = "agent.server_active_async_calls";
+const std::string kTensorPipeRPCTimeoutErrorStr =
```
nit: kTensorPipeRpcTimeoutErrorStr

lw replied:
nit: no need for the `TensorPipe` prefix, this is in anonymous namespace, it's local to this file

osalpekar commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -418,7 +421,8 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
       auto& timeoutFuturesVector = timeoutMap_[expirationTime];
-      timeoutFuturesVector.emplace_back(futureResponseMessage);
+      timeoutFuturesVector.emplace_back(
+          std::make_pair(futureResponseMessage, timeout));
```
I think you might not need to call `std::make_pair` with `emplace_back`?

rohan-varma replied:
Great catch!

osalpekar commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2833,7 +2833,6 @@ def test_return_future(self):
             )
 
-    @_skip_if_tensorpipe_agent
     @dist_init
```
Has this been stress-tested? Just checking to ensure we don't introduce more flaky TensorPipe agent tests

rohan-varma replied:
Ran it with 500 stresstests and they all passed.

lw reviewed (approved):
Looks great, thanks!

rohan-varma committed:
Update on "Enable RRef timeout for tensorpipe"


Enables RRef timeout support in TP agent by having TP agent mark
timeout errors with `makeRPCError` API. Also does some refactoring so TP agent
can print out the timeout for each future that has timed out.

Differential Revision: [D21881475](https://our.internmc.facebook.com/intern/diff/D21881475/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enable RRef timeout for tensorpipe"


Enables RRef timeout support in TP agent by having TP agent mark
timeout errors with `makeRPCError` API. Also does some refactoring so TP agent
can print out the timeout for each future that has timed out.

Differential Revision: [D21881475](https://our.internmc.facebook.com/intern/diff/D21881475/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enable RRef timeout for tensorpipe"


Enables RRef timeout support in TP agent by having TP agent mark
timeout errors with `makeRPCError` API. Also does some refactoring so TP agent
can print out the timeout for each future that has timed out.

Differential Revision: [D21881475](https://our.internmc.facebook.com/intern/diff/D21881475/)

[ghstack-poisoned]
